source/Patches/EndGameSummary.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using UnityEngine;
+using TownOfUs.Roles;
+using TownOfUs.Roles.Modifiers;
+using HarmonyLib;
+
+namespace TownOfUs.Patches
+{
+    [HarmonyPatch]
+    static class EndGameSummary
+    {
+        static class AdditionalTempData
+        {
+            public static List<PlayerInfo> playerData = new List<PlayerInfo>();
+
+            //leaving this in as i assume it maybe needed in future, feel free to remove if you think otherwise.
+            public static void clear()
+            {
+                playerData.Clear();
+            }
+
+            internal class PlayerInfo
+            {
+                public byte PlayerId { get; set; }
+                public string PlayerName { get; set; }
+                public Faction Faction { get; set; }
+                public Role CurrentRole { get; set; }
+                public Modifier Modifier { get; set; }
+                public int TasksCompleted { get; set; }
+                public int TasksTotal { get; set; }
+            }
+        }
+
+        public static void UpdatePlayerInfo()
+        {
+            Il2CppSystem.Collections.Generic.List<GameData.PlayerInfo> players = GameData.Instance.AllPlayers;
+
+            var pinfolist = new List<AdditionalTempData.PlayerInfo>();
+
+            foreach (GameData.PlayerInfo player in players)
+            {
+                byte pid = player.PlayerId;
+                string name = player.PlayerName;
+                Role role = Role.GetRole(player);
+                Modifier modifier = Modifier.GetModifier(player);
+                Faction faction = role.Faction;
+                int tasksdone = 0;
+                int taskstotal = 0;
+
+                AdditionalTempData.PlayerInfo pinfo = new AdditionalTempData.PlayerInfo();
+                pinfo.PlayerName = name;
+                pinfo.CurrentRole = role;
+                pinfo.Faction = faction;
+                pinfo.Modifier = modifier;
+
+                foreach (var task in player.Tasks)
+                {
+                        if (task.Complete == true)
+                        {
+                            tasksdone++;
+                        }
+                        taskstotal++;
+                }
+                
+                pinfo.TasksCompleted = tasksdone;
+                pinfo.TasksTotal = taskstotal;
+                pinfolist.Add(pinfo);
+            }
+            AdditionalTempData.playerData = pinfolist;
+        }
+
+        public static bool HasTasks(Faction faction)
+        {
+            return faction == Faction.Crewmates;
+        }
+
+        static bool IsNull(AdditionalTempData.PlayerInfo pinfo, string type)
```

Comment:
Suggestion: I think this method will cause more confusion in the long run than it will save you time here. Especially because "Role" and "Mod" are strings. For example, it's not obvious that you need to put "Mod" instead of "Modifier", and there's not really any feedback to know you've gotten it wrong.

Reply:
i am well aware i was doing an ugly way of doing this, but i found a big cause of problem for me was the null exceptions from modifier mostly, but role also. So once i got it working, i did not review it, after making the other changes you suggested, i will look at improving this.

Reply:
it seems !=null works now.. baffling, must have been something else in my code that was breaking it before, either way, simplified this now to if(variable!=null)